### ISSUE TITLE
lxc-autostart never exits when called non-interactively

### ISSUE THREAD

**Issue by jzacsh** [lxc/lxc#1681]
_Note: I moved "required information" section to the bottom._

# Issue description

**tl;dr** `lxc-autostart` hangs indefinitely (but seems to do stuff) as an `ExecStart` systemd service.

I believe something about the lxc-autostart command is detecting having a tty
(maybe?) and not exiting when I run it via my own systemd user unit file in
`$HOME/.config/systemd/user/lxc-autostart.service` _(pasted below)_. This is in contrast to
running `lxc-autostart` from my shell the program exiting (with `0`) instantly
_(see "steps" below)_.

```
[Unit]
Description=Autostart Unpriveleged LXCs

[Service]
Type=simple
ExecStart=/bin/bash -c 'lxc-autostart'
#TimeoutStopSec=3min 30s
# toggle this ^ timeout to see this controls precisely how long until containers are turned
# off and `systemctl status` of this unit reports faiure due to time-out.

[Install]
WantedBy=default.target
```

I'm guessing this is a duplicate or at least related to #657, but I thought maybe my issue is more distinctly just that lxc-autostart is behaving unexpectedly (at least, I don't see anything in the man page about not exiting).

# Steps to reproduce

 1. `lxc-ls -f` to ensure your containers are STOPPED
 2. establish baseline behavior of `lxc-autostart` on your systemd:
```
$ time ( lxc-autostart ; )

real    0m0.010s
user    0m0.008s
sys     0m0.000s
```
 3. try starting above-pasted unit file and watch its status:
```
$ systemctl --user start lxc-autostart.service
$ journalctl --follow --pager-end --catalog --user-unit lxc-autostart.service
# everything's great....
# until TimeoutStopSec (or check systemctl --user show lxc-autostart.service)
```

You can also change `Type=oneshot` to see lxc-autostart hanging your `systemctl
start`.

NOTE: Just using the direct path to `lxc-autostart`, in `ExecStart`, doesn't change behavior.

# Required information

 * Distribution: Ubuntu
 * Distribution version: 16.04.2
 * The output of
   * `lxc-start --version` is `2.0.8`
   * `lxc-checkconfig`
```
Kernel configuration not found at /proc/config.gz; searching...
Kernel configuration found at /boot/config-4.4.0-83-generic
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Warning: newuidmap is not setuid-root
Warning: newgidmap is not setuid-root
Network namespace: enabled
Multiple /dev/pts instances: enabled

--- Control groups ---
Cgroup: enabled
Cgroup clone_children flag: enabled
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: enabled
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled
Macvlan: enabled
Vlan: enabled
Bridges: enabled
Advanced netfilter: enabled
CONFIG_NF_NAT_IPV4: enabled
CONFIG_NF_NAT_IPV6: enabled
CONFIG_IP_NF_TARGET_MASQUERADE: enabled
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled
FUSE (for use with lxcfs): enabled

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: enabled
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: enabled
File capabilities: enabled

Note : Before booting a new kernel, you can check its configuration
usage : CONFIG=/path/to/config /usr/bin/lxc-checkconfig
```
   * `uname -a` is `Linux jznuc 4.4.0-83-generic #106-Ubuntu SMP Mon Jun 26 17:54:43 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
   * `cat /proc/self/cgroup`
```
11:cpu,cpuacct:/user.slice
10:hugetlb:/
9:perf_event:/
8:memory:/user/jzacsh/0
7:cpuset:/
6:pids:/user.slice/user-1000.slice
5:devices:/user.slice
4:net_cls,net_prio:/
3:freezer:/user/jzacsh/0
2:blkio:/user.slice
1:name=systemd:/user.slice/user-1000.slice/session-c2.scope
```
   * `cat /proc/1/mounts`
```
sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
udev /dev devtmpfs rw,nosuid,relatime,size=8145188k,nr_inodes=2036297,mode=755 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
tmpfs /run tmpfs rw,nosuid,noexec,relatime,size=1633120k,mode=755 0 0
/dev/sda2 / ext4 rw,relatime,errors=remount-ro,data=ordered 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev 0 0
tmpfs /run/lock tmpfs rw,nosuid,nodev,noexec,relatime,size=5120k 0 0
tmpfs /sys/fs/cgroup tmpfs rw,mode=755 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/lib/systemd/systemd-cgroups-agent,name=systemd 0 0
pstore /sys/fs/pstore pstore rw,nosuid,nodev,noexec,relatime 0 0
efivarfs /sys/firmware/efi/efivars efivarfs rw,nosuid,nodev,noexec,relatime 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids,release_agent=/run/cgmanager/agents/cgm-release-agent.pids 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset,clone_children 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event,release_agent=/run/cgmanager/agents/cgm-release-agent.perf_event 0 0
cgroup /sys/fs/cgroup/hugetlb cgroup rw,nosuid,nodev,noexec,relatime,hugetlb,release_agent=/run/cgmanager/agents/cgm-release-agent.hugetlb 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=23,pgrp=1,timeout=0,minproto=5,maxproto=5,direct 0 0
hugetlbfs /dev/hugepages hugetlbfs rw,relatime 0 0
debugfs /sys/kernel/debug debugfs rw,relatime 0 0
mqueue /dev/mqueue mqueue rw,relatime 0 0
fusectl /sys/fs/fuse/connections fusectl rw,relatime 0 0
none /run/shm tmpfs rw,nosuid,nodev,noexec,relatime 0 0
/dev/sda1 /boot/efi vfat rw,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro 0 0
binfmt_misc /proc/sys/fs/binfmt_misc binfmt_misc rw,relatime 0 0
/dev/sdc1 /media/terrahome ext4 rw,relatime,data=ordered 0 0
/dev/sdb1 /media/terrasquirrel/hippomedia ext4 rw,relatime,data=ordered 0 0
cgmfs /run/cgmanager/fs tmpfs rw,relatime,size=100k,mode=755 0 0
lxcfs /var/lib/lxcfs fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
tmpfs /run/user/1000 tmpfs rw,nosuid,nodev,relatime,size=1633120k,mode=700,uid=1000,gid=1000 0 0
gvfsd-fuse /run/user/1000/gvfs fuse.gvfsd-fuse rw,nosuid,nodev,relatime,user_id=1000,group_id=1000 0 0
```

# Information to attach

 - [] any relevant kernel output (`dmesg`): no, but pasting anyway
```
[12663.986978] IPv6: ADDRCONF(NETDEV_UP): vethHUWAAE: link is not ready
[12663.987360] device vethHUWAAE entered promiscuous mode
[12664.022943] eth0: renamed from vethHUWAAEp
[12664.052133] IPv6: ADDRCONF(NETDEV_CHANGE): vethHUWAAE: link becomes ready
[12664.052329] lxcbr0: port 1(vethHUWAAE) entered forwarding state
[12664.052370] lxcbr0: port 1(vethHUWAAE) entered forwarding state
[12679.078562] lxcbr0: port 1(vethHUWAAE) entered forwarding state
```
 - [ ] container log (The <log> file from running `lxc-start -n <c> -l <log> -o DEBUG`)
 - [x] the containers configuration file
```
# Template used to create this container: /usr/share/lxc/templates/lxc-download
# Parameters passed to the template: -d ubuntu -r xenial -a amd64
# Template script checksum (SHA-1): 9748088977ba845f625e45659f305a5395c2dc7b
# For additional config options, please look at lxc.container.conf(5)

# Uncomment the following line to support nesting containers:
#lxc.include = /usr/share/lxc/config/nesting.conf
# (Be aware this has security implications)


# Distribution configuration
lxc.include = /usr/share/lxc/config/ubuntu.common.conf
lxc.include = /usr/share/lxc/config/ubuntu.userns.conf
lxc.arch = x86_64

# Container specific configuration
lxc.id_map = u 0 100000 65536
lxc.id_map = g 0 100000 65536
lxc.rootfs = /home/jzacsh/.local/share/lxc/lanwebdeskgo/rootfs
lxc.rootfs.backend = dir
lxc.utsname = lanwebdeskgo

# Network configuration
lxc.network.type = veth
lxc.network.link = lxcbr0
lxc.network.flags = up
lxc.network.hwaddr = 00:16:3e:9f:a2:fa

# jzacsh@ 2017-07-01; autostart this container
lxc.start.auto = 1
```

**Comment by jzacsh** [comment]
I should clarify that I can ping and curl things out of my container _(which starts a little simpleserver)_ in *both* cases (the shell call to `lxc-autostart` and the systemctl call).

**Comment by brauner** [comment]
@stgraber you know anything about this?

**Comment by jzacsh** [comment]
i'd be curious if others can even reproduce this, or if it's just me.

If it's just me, I'll try diffing an strace of the two cases' executions, to see if anything obviously helpful pops out.

**Comment by stvnjacobs** [comment]
@jzacsh I can reproduce this.

Occasionally, the `systemctl --user start lxc-autostart.service` finishes successfully, but a large majority of the time it fails due to a timeout and kills the started containers along with it.  I am calling `/usr/bin/lxc-autostart` directly, as you noted that you also tried.  I have also tried both `simple` and `oneshot` for `ExecStart`.

I am now using the following for `ExecStart`, but have been unable to determine any difference between when a timeout will occur and when it will exit correctly.

```
ExecStart=/usr/bin/lxc-autostart -l trace -o /tmp/lxc-autostart.log
```

The container always starts successfully, it is just a matter of whether it will be killed by a timeout.  

I am available to provide any additional logs and/or details, if needed.

- `Linux stj 4.9.0-3-amd64 #1 SMP Debian 4.9.30-2+deb9u2 (2017-06-26) x86_64 GNU/Linux`
- All checked configs are enabled in `lxc-checkconfig`
- `lxc-start` version is `2.0.7`

**Comment by pjanouch** [comment]
Can reproduce on Debian Stretch with a Debian Stretch LXC unprivileged container.

**Comment by juppin** [comment]
My solution to start my unprivileged containers of user container:

Create executable bash script /root/start_container.sh
`
#!/bin/bash
/bin/su -l container << @EOF
/usr/bin/lxc-autostart
exit
@EOF
exit 0
`

Start it with crontab at reboot or i think systemd should also work
@reboot sleep 15 && /root/start_container.sh
